### DISCUSSION TITLE
Fixed size buffers per thread

### DISCUSSION THREAD

**Discussion by ishaanshah** [mitsuba-renderer/mitsuba3#817] [Q&A]
Hi,

Is it possible to create a fixed-size buffer for each thread in the rendering loop? Something like `mi.VectorXf`, where the `X` is defined during JIT compilation based upon some variable.
I am trying to implement [Efficient Rendering of Layered Materials using an Atomic Decomposition with Statistical Operators [Belcour, 2018] ](https://belcour.github.io/blog/research/publication/2018/05/05/brdf-realtime-layered.html) in Mitsuba 3. The paper computes a set of `N` energies and variances for `N` layers depending on the shading point and viewing angle. Hence I need a way of storing these `N` floats per thread for further evaluation and sampling. Currently, I can render at max four layers as the floats can be stored in a `Vector4f`. However, it would be good to have an implementation that works for an arbitrary number of layers.

Thanks

**Comment by rtabbara** [comment]
Hi @ishaanshah ,

Just taking a look at the accompanying (old) Mitsuba code from the paper, I think you'd similarly want to store your parameters as an N-sized vector/list of `mi.Float` variables
